### DISCUSSION TITLE
Candlestick patterns

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#220] [Ideas]
[Candlestick patterns](https://en.wikipedia.org/wiki/Candlestick_pattern) are a unique form of price analysis.  I'm opening this discussion to get feedback and to discuss development issues on this category of indicators.  We're adding these to v1.x.  Thoughts?

**Comment by GmiSoft** [comment]
Hi, 

I started exploring Candlestick patterns and it's very useful for trading and it's helpful to identify trading opportunities.
I'm trying to detect some patterns like Hammer or Morning star and the biggest constraint I have encountered is how to define thresholds: For example, based on High/Low Open Close, When do I consider a candle as a candlestick with a big Body or a Doji or spinning.

![image](https://user-images.githubusercontent.com/21221673/101809547-fd9acf80-3b17-11eb-92ff-05ee1043ceca.png)

For a long shadow, we have a formula of 2/3 or more of the total range but not for the Hammer.

I think it will be very useful to have dynamic configurations (with default values) for these thresholds and the user chooses values based on his requirements.

 I hope again that I'm not out of the box.

**Reply by DaveSkender** [reply]
There are definitely some rules for these, for example, [Hanging Man](https://en.wikipedia.org/wiki/Hanging_man_(candlestick_pattern)) specifies that

>In order for a candle to be a valid hanging man most traders say the lower wick must be two times greater than the size of the body portion of the candle, and the body of the candle must be at the upper end of the trading range.

Or at least most do.  And you're right, we'll likely need to parameterize some of these to thresholds, but provide a default setting.  Patterns like double-top and cup and handle will be a bit more tricky to analyze programmatically though.

**Comment by GmiSoft** [comment]
Hi Dave,

I take a look at the TA-Lib C++ project and I found this piece of code. And I just wanted to share this with the community, because I was a little surprised.
I think, using the average of the last 10th candles to define current candle properties not enough, and maybe not secure to detect patterns.

What do you think?

```C 
const TA_CandleSetting TA_CandleDefaultSettings[] = {
        /* real body is long when it's longer than the average of the 10 previous candles' real body */
        { TA_BodyLong, TA_RangeType_RealBody, 10, 1.0 },
        /* real body is very long when it's longer than 3 times the average of the 10 previous candles' real body */
        { TA_BodyVeryLong, TA_RangeType_RealBody, 10, 3.0 },
        /* real body is short when it's shorter than the average of the 10 previous candles' real bodies */
        { TA_BodyShort, TA_RangeType_RealBody, 10, 1.0 },
        /* real body is like doji's body when it's shorter than 10% the average of the 10 previous candles' high-low range */
        { TA_BodyDoji, TA_RangeType_HighLow, 10, 0.1 },
        /* shadow is long when it's longer than the real body */
        { TA_ShadowLong, TA_RangeType_RealBody, 0, 1.0 },
        /* shadow is very long when it's longer than 2 times the real body */
        { TA_ShadowVeryLong, TA_RangeType_RealBody, 0, 2.0 },
        /* shadow is short when it's shorter than half the average of the 10 previous candles' sum of shadows */
        { TA_ShadowShort, TA_RangeType_Shadows, 10, 1.0 },
        /* shadow is very short when it's shorter than 10% the average of the 10 previous candles' high-low range */
        { TA_ShadowVeryShort, TA_RangeType_HighLow, 10, 0.1 },
        /* when measuring distance between parts of candles or width of gaps */
        /* "near" means "<= 20% of the average of the 5 previous candles' high-low range" */
        { TA_Near, TA_RangeType_HighLow, 5, 0.2 },
        /* when measuring distance between parts of candles or width of gaps */
        /* "far" means ">= 60% of the average of the 5 previous candles' high-low range" */
        { TA_Far, TA_RangeType_HighLow, 5, 0.6 },
        /* when measuring distance between parts of candles or width of gaps */
        /* "equal" means "<= 5% of the average of the 5 previous candles' high-low range" */
        { TA_Equal, TA_RangeType_HighLow, 5, 0.05 }
    };
```

**Reply by DaveSkender** [reply]
Lots of thoughts here.

- **Consistency principle** - as long as an indicator can consistently identify a market behavior, its theoretical accuracy is less important.
- **[Bygone Era](https://www.merriam-webster.com/dictionary/bygone)** - older data rots to a point where it is no longer representative of current periods.  Going back too far would include stale unrepresentative data.
- **TA-Lib** - is an old library that uses older techniques for programming.  I'm not sure when it was first developed, but it's probably 25+ years old.  It may have been developed to accommodate memory constraints where a smaller lookback period was forced.
- **Intent** - in thinking of the [Long Body Pattern](https://school.stockcharts.com/doku.php?id=chart_analysis:candlestick_pattern_dictionary#long_body_long_day), we're really looking for a change in _current_ behavior, not necessarily a statistically accurate size definition over all of the time.  What was considered a large candle last year might not be considered big now, especially if you think about how [systemic volatility](https://www.investopedia.com/terms/s/systematicrisk.asp) changes over time.

**Reply by DaveSkender** [reply]
Thinking about this from a statistics view, what is "long" in a Long Body Pattern?  I think "longer than average".  Then I think: longer than X deviations from the mean size, where [mean and standard deviation](https://en.wikipedia.org/wiki/Standard_deviation) are calculated from a [statistically significant sample size](https://www.qualtrics.com/experience-management/research/determine-sample-size).  :neckbeard:

**Reply by DaveSkender** [reply]
The question you can answer, through analysis: does a 10-period sample have the same mean and standard deviation body size as a much larger sample?  

Shameless advertisement: you can use the [Standard Deviation](https://daveskender.github.io/Stock.Indicators/indicators/StdDev/#content) indicator in this library for your analysis, though you'd have to make a synthetic `Quote` where Close = Body Size.  See [side note on GetStdDev](https://github.com/DaveSkender/Stock.Indicators/discussions/239#discussioncomment-230241).

**Comment by vebollu** [comment]
https://csharp.hotexamples.com/examples/-/ATArray/-/php-atarray-class-examples.html

With the help of above code I made changes as per my requirement.

**Comment by vebollu** [comment]
this is what I have in my current system, if you think this is worth please include this in your code.

                bool hammer, invertedHammer, hangingMan, shootingStar, bullishEngulfing, bullishHarami, doji;
                GetCandle(quotes[i], out hammer, out invertedHammer, out hangingMan, out shootingStar, out doji);
                GetCandle(quotes[i], quotes[i - 1], out bullishEngulfing, out bullishHarami);

        private static void GetCandle(Quote curr, out bool hammer, out bool invertedHammer, out bool hangingMan, out bool shootingStar, out bool doji)
        {
            //var curr = p.Last();
            var High = Convert.ToSingle(curr.High);
            var Low = Convert.ToSingle(curr.Low);
            var Open = Convert.ToSingle(curr.Open);
            var Close = Convert.ToSingle(curr.Close);

            hammer = (((High - Low) > 3 * (Open - Close)) & ((Close - Low) / (.001f + High - Low) > 0.6f) & ((Open - Low) / (.001f + High - Low) > 0.6f));
            invertedHammer = (((High - Low) > 3 * (Open - Close)) & ((High - Close) / (0.001f + High - Low) > 0.6f) & ((High - Open) / (0.001f + High - Low) > 0.6f));
            hangingMan = (((High - Low) > 4 * (Open - Close)) & ((Close - Low) / (.001f + High - Low) >= 0.75f) & ((Open - Low) / (.001f + High - Low) >= 0.75f));
            shootingStar = (((High - Low) > 4 * (Open - Close)) & ((High - Close) / (0.001f + High - Low) >= 0.75f) & ((High - Open) / (0.001f + High - Low) >= 0.75f));
            doji = Open == Close | (Math.Abs(Open - Close) <= ((High - Low) * 0.1f));

        }

        private static void GetCandle(Quote curr, Quote prev, out bool bullishEngulfing, out bool bullishHarami)
        {
           // var High = Convert.ToSingle(curr.High);
           // var Low = Convert.ToSingle(curr.Low);
            var Open = Convert.ToSingle(curr.Open);
            var Close = Convert.ToSingle(curr.Close);

            var O1 = Convert.ToSingle(prev.Open);
            var C1 = Convert.ToSingle(Convert.ToSingle(prev.Close));
            // var BullishEngulfing = ((O1 > C1) & (Close > Open) & (Close >= O1) & (C1 >= Open) & ((Close - Open) > (O1 - C1)));
            bullishEngulfing = ((O1 > C1) & (Close > Open) & (Close >= O1) & (C1 >= Open) & ((Close - Open) > (O1 - C1)));
            bullishHarami = ((O1 > C1) & (Close > Open) & (Close <= O1) & (C1 <= Open) & ((Close - Open) < (O1 - C1)));

        }

**Comment by DaveSkender** [comment]
Just as an update, @metathron is working on a baseline set of candlestick patterns.  Once we have a baseline approach established, we'll be able to add more incrementally.

**Reply by DaveSkender** [reply]
@metathron, I know you're busy with your day-job, I'd like to start implementing these one at a time -- any chance you're able to contribute to the library?  If not, I'd like to start implementing myself, but wanted to give you first choice.  I have a few lined up in [the backlog](https://github.com/DaveSkender/Stock.Indicators/projects/1?card_filter_query=%40metathron) with your name on them.

**Comment by DaveSkender** [comment]
I'm going to incrementally start working on these now.  There are quite a few to consider and would love some help on these if anyone is interested in contributing.  To start, I'll do a few of these to establish a baseline approach that anyone can follow.

I'll also put these in [the Backlog](https://github.com/users/DaveSkender/projects/1) in a prioritized manner, starting with the simpler ones.  I don't suspect that I'll do the "complex" ones, but will take a look after the easier patterns are implemented.

If you want to contribute:
1. Mention which item you will work on in this discussion
2. I'll convert [the Backlog](https://github.com/users/DaveSkender/projects/1) note to an issue for focused discussion and tracking
3. Follow the [contributing guidelines](https://daveskender.github.io/Stock.Indicators/docs/CONTRIBUTING.html)

I'm pulling recipes from these publicly available sources:
- [Zerodha](https://zerodha.com/varsity/chapter/getting-started-candlesticks)
- [StockCharts.com](https://school.stockcharts.com/doku.php?id=chart_analysis:chart_patterns)
- [TA-Lib](https://ta-lib.org/function.html)
- [TickBlaze](https://tickblaze.com/scripts/pattern), as a backup

```bash
# Single Candlestick patterns:
Marubozu
Bullish Marubozu
Bearish Marubozu
Doji
Spinning Tops
Paper umbrella
Hammer
Hanging man
Shooting star

# Combination patterns:
Engulfing pattern
Bullish Engulfing
Bearish Engulfing
Harami
Bullish Harami
Bearish Harami
Piercing Pattern
Dark cloud cover
Morning Star
Evening Star

# Complex Reversal patterns:
Double Top Reversal
Double Bottom Reversal
Head and Shoulders Top
Head and Shoulders Bottom
Falling Wedge
Rising Wedge
Rounding Bottom
Triple Top Reversal
Triple Bottom Reversal
Bump and Run Reversal

# Complex Continuation patterns:
Flag, Pennant
Symmetrical Triangle
Ascending Triangle
Descending Triangle
Rectangle
Price Channel
Measured Move - Bullish
Measured Move - Bearish
Cup with Handle

# TA-LIB legacy
CDL2CROWS           Two Crows
CDL3BLACKCROWS      Three Black Crows
CDL3INSIDE          Three Inside Up/Down
CDL3LINESTRIKE      Three-Line Strike 
CDL3OUTSIDE         Three Outside Up/Down
CDL3STARSINSOUTH    Three Stars In The South
CDL3WHITESOLDIERS   Three Advancing White Soldiers
CDLABANDONEDBABY    Abandoned Baby
CDLADVANCEBLOCK     Advance Block
CDLBELTHOLD         Belt-hold
CDLBREAKAWAY        Breakaway
CDLCLOSINGMARUBOZU  Closing Marubozu
CDLCONCEALBABYSWALL Concealing Baby Swallow
CDLCOUNTERATTACK    Counterattack
CDLDARKCLOUDCOVER   Dark Cloud Cover
CDLDOJI             Doji
CDLDOJISTAR         Doji Star
CDLDRAGONFLYDOJI    Dragonfly Doji
CDLENGULFING        Engulfing Pattern
CDLEVENINGDOJISTAR  Evening Doji Star
CDLEVENINGSTAR      Evening Star
CDLGAPSIDESIDEWHITE Up/Down-gap side-by-side white lines
CDLGRAVESTONEDOJI   Gravestone Doji
CDLHAMMER           Hammer
CDLHANGINGMAN       Hanging Man
CDLHARAMI           Harami Pattern
CDLHARAMICROSS      Harami Cross Pattern
CDLHIGHWAVE         High-Wave Candle
CDLHIKKAKE          Hikkake Pattern
CDLHIKKAKEMOD       Modified Hikkake Pattern
CDLHOMINGPIGEON     Homing Pigeon
CDLIDENTICAL3CROWS  Identical Three Crows
CDLINNECK           In-Neck Pattern
CDLINVERTEDHAMMER   Inverted Hammer
CDLKICKING          Kicking
CDLKICKINGBYLENGTH  Kicking - bull/bear determined by the longer marubozu
CDLLADDERBOTTOM     Ladder Bottom
CDLLONGLEGGEDDOJI   Long Legged Doji
CDLLONGLINE         Long Line Candle
CDLMARUBOZU         Marubozu
CDLMATCHINGLOW      Matching Low
CDLMATHOLD          Mat Hold
CDLMORNINGDOJISTAR  Morning Doji Star
CDLMORNINGSTAR      Morning Star
CDLONNECK           On-Neck Pattern
CDLPIERCING         Piercing Pattern
CDLRICKSHAWMAN      Rickshaw Man
CDLRISEFALL3METHODS Rising/Falling Three Methods
CDLSEPARATINGLINES  Separating Lines
CDLSHOOTINGSTAR     Shooting Star
CDLSHORTLINE        Short Line Candle
CDLSPINNINGTOP      Spinning Top
CDLSTALLEDPATTERN   Stalled Pattern
CDLSTICKSANDWICH    Stick Sandwich
CDLTAKURI           Takuri (Dragonfly Doji with very long lower shadow)
CDLTASUKIGAP        Tasuki Gap
CDLTHRUSTING        Thrusting Pattern
CDLTRISTAR          Tristar Pattern
CDLUNIQUE3RIVER     Unique 3 River
CDLUPSIDEGAP2CROWS  Upside Gap Two Crows
CDLXSIDEGAP3METHODS Upside/Downside Gap Three Methods
```

**Comment by jimtollan** [comment]
Hi Dave. regards the TA-LIB and the exisiting Candlestick signal/identification. 

Would it be REALLY bad form to create a wrapper over TA-LIB and use a Quotes extension method to use both libraries in conjunction?? My thinking here is that the TA-LIB functions could be consumed by the extension methods in the 1st instance and then, as time passes/collaboration, the implementations could be replaced with native Skender functionality. 

Below is a basic example that I was attempting to use to highlight the approach:

![image](https://user-images.githubusercontent.com/87867061/148693780-d3c05fb0-0c9b-4f26-9b65-75fb831b4c55.png)

CandleSignal class for now purely mimics the TA-LIB properties in name. Obviously, that would be formalised into a Skender coding standard.

![image](https://user-images.githubusercontent.com/87867061/148693792-4e38c178-1cf8-466a-938e-ea5fd7e81c80.png)

usage: 

`CandleSignal` candleSignal = quotes.GetCandleDragonflyDoji(startIndex, endIndex);

TBH - i don't fully understand the TA-LIB implementation as the docs are quite thin, also, my method is of course lacking in lots of defensive checks - but hopefully you get the gist.

**Reply by DaveSkender** [reply]
> Would it be REALLY bad form to create a wrapper over TA-LIB

I had used TA-Lib exclusively in my trading apps for about 10 years, so I'm quite familiar with it.  It's not easy to use IMO.  Because it's so old and no longer maintained, it's not directly compatible with modern .NET anymore, so we'd have to use an unusual integration, a wrapper, or someone's ported rewrite (there are quite a few of these available).  In other words, it'd have to be a wrapper of wrappers (this TACore you're showing is likely just another wrapper).

To answer your question directly, yes, I think it's bad form to just wrap the old TA-Lib here in this library, but I do appreciate the "do the fast thing", then replace it with the better thing concept.  I'm also trying to avoid using _any_ third-party libraries.  With that said, TA-Lib is a great legacy and a good place for inspiration.

**Reply by DaveSkender** [reply]
And pragmatically speaking, it's probably almost as fast to build new as it is to wrap each of these indicators anyway.

**Reply by jimtollan** [reply]
cheers again dave. i found the c# TA-LIB on gitHub and it does look quite convoluted tbh. here's the function for the above:

https://github.com/DavidFuchs42/TALibraryInCSharp/blob/master/TALibraryInCSharp/TAFunc/TA_CdlDragonflyDoji.cs

Quick a mouthful to swallow. looks like it was decomplied from something else imho -based on the labels and variables in use. But yeah, the idea of porting against a 3rd party isn't the best option, but may be a good transitional approach.

**Reply by DaveSkender** [reply]
Omg, that code! Wow.  This is a good example of why clean original work is needed.

**Reply by jimtollan** [reply]
yes, jaw dropping!! I gave up trying to figure what it was doing tbh. As there is NO documentation or usage examples, i was planning to create a number of FAKE patterns that should be triggered by the signals and simple explore the outputs to figure out how it can be used. I guess in the longer run, having fake sets that match (and don't match) the particular function would be useful for integration tests etc.

And yup - for sure clean original work is the way fwd. i guess my grit is purely down to looking to _use it today_ type scenario.

**Reply by DaveSkender** [reply]
Yes, test data and TDD is a great place to start!  If you're interested in contributing test data or test elements to this library, please do!  I'm always looking for some help building, even if it's just parts and pieces, code reviews, feedback, ideas, etc.

**Reply by jimtollan** [reply]
dave - i can certainly mashup some json test **Quote** data that could be used against each function (positive and negative paths). likewise, can be on hand to test new functionality and provide additional ideas/feedback. to arrive at this data. 

i shall of course, in the interim, do the _dirty_ and wrap the TA-LIB with the required extension methods. in truth, i could then use those to A/B check any new functionality as it becomes available.

**Reply by jimtollan** [reply]
in fact, i may build a small GUI that allows me to visually design the required patterns and save to json. That way, I can quickly put together the patterns as they appear to the eye (and thus can create a number of sets per candlestick type to test tolerances etc)... just thinking out loud -may not happen :D

**Reply by jimtollan** [reply]
FYI - i found an implementation for .net core which seems to have a less crappy codebase for the candles. still looks a bit *functional* but is easier to grok. using the same candle recognition example from yesterday:

https://github.com/hmG3/TA-Lib.NETCore/blob/master/src/TALib.NETCore/TACdl/TA_CdlDragonflyDoji.cs

hope this helps in some small way moving fwd.

**[edit]** also, FYI. when using the candles functions with Bitcoin, i had to truncate the decimals from the price, otherwise, it gave me very few candidates (due to exact comparisons on price). this could be an issue when working with crypto based candles as many of the coins come with a large number of decimals. 

Also, I'm not a great maths guy, but do wonder if an absolute value comparison is the BEST way to compare the OHLC values. To me, it seems like a better approach would be to *box* the incoming values into a defined minimum and maximum range for the candle attributes (i.e. 1-100 based across the entire set). That way, we don't get weird values for BTC (price 49234.456) vs ONE (0.2999345) etc. It also means that we get truer representations of the long legged doji type occurences vs spinning top etc, etc.

Hope this makes sense - You'll know this stuff better than me, but this type of common currency approach seems key to me (I'll try creating a small method that tkes the Quotes object and Normalises/smooths all the output arrays into a common currency and see if we get more hits and better consistancy on the output candidates)

**Reply by jimtollan** [reply]
So, to clarify my *Common Currency* thinking. If the OHLC range was representing 30 BTC candles, we'd look for the highest and lowest value on any OHLC component, across the range. From that, we'd determine that our highest value represented 100 and the lowest value, 1. we'd then pro-rate all the other values using the upper/lower found values.

just a thought - feel free to laugh at the niavity :D

**Reply by DaveSkender** [reply]
I tend to like to find source information on these (e.g. academic or other documentation from original authors), so I'll be doing additional research before just reading someone else's code and hoping it's right.

Regarding precision, I tend to use `decimal` precision for candle based stuff so we should be okay.  I suspect the old TA-Lib was trying to use single `float` precision.  They would have never predicted cryptocurrencies when they wrote it.  I've had a whole internal monologue on this topic: see #649

**Reply by jimtollan** [reply]
yup -i think the fact that crypto wasn't around back in the TA-Lib inception days is what's causing the issues here. definitely the `decimal value is the way to go. as it stands, there's no value in wrapping the TA-Lib, even the version that uses .net core.

will be around as a sounding board as/when you get into this in anger... all the best again

**Reply by jimtollan** [reply]
btw - prorating the quotes from 0-100 on 2 decimal places SEEMS to work well across a range of cryptos on different price scales. here's an example using BTC:

Normal Scale:
![image](https://user-images.githubusercontent.com/87867061/148865154-6465163d-77bf-455a-bbee-d8963b50ae0f.png)

Pro-rated (0-100)
![image](https://user-images.githubusercontent.com/87867061/148865687-5e24d1a5-23c9-4cb6-afbc-71fb5d722107.png)

Just tinkering with this idea and pasting it here to see if the wisdom of the crowd sees any major flaws in this approach.

**Comment by DaveSkender** [comment]
Has anyone taken a look at our implementation of the [Marubozu](https://daveskender.github.io/Stock.Indicators/indicators/Marubozu/#content) candlestick pattern?  Feedback appreciated.  I may start looking at implementing these again soon but wanted to make sure we have a good baseline approach before proceeding.

@metathron, I've been holding a bunch for you, but it's been a while, so I'll be releasing that hold soon for the sake of progress.

**Reply by jimtollan** [reply]
Hi Dave. Yes, I took a look at the Marobozu example and have a few comments (that may or may not be in keeping with your thinking). I wondered if rather than returning a boolean flag (isbullish etc), it may be better to return an object that returned a wider range of values per identification. This could comprise of an enum value (Bullish, Bearish, Neutral, None, etc) and a percentage value representing the actual match found plus the existing decimal price as per the current result. That way, you could examine the payload based on looking for a particular Enum value, as well as reviewing their percentage (strength) etc. A non matching candle would simply return None for the Enum member.

This may violate your intention, but was thinking in this way as a precursor to how you'd want to return objects for multi-candle patterns and an enum could be used to represent a wider range of *opinion* on the candle(s).

**Reply by jimtollan** [reply]
kinda thinking out loud here now, so take with a pinch of salt. potentially, for multi candles, a sparse structure could be returned which simply identified the start and end indexes of the identification, as well as the strength of the recognition (against some baseline). whilst this wouldn't match up as a one-to-one tally with the quotes list, it would identify the groupings (if found) and allow for a simple analysis of them.

I'm also imagining being able to fluently wire up a number of patterns in a chain and have a return value that contained the location and strngth of any such occurrences. however, that's pushing ahead of the base requirements. for now, a simple single pattern identification using the sparse return would work well i think.

ok, end of thinking out loud!! :)

**Reply by DaveSkender** [reply]
I see the old TA-Lib only produces one integer value as its output for candles:
```
+200 bullish with confirmation
+100 bullish (most commonly used)
   0 no pattern matched
-100 bearish
-200 bearish with confirmation
```
I think we can do better.  @jimtollan, I like your idea of `enum`, which is really just a fancy `integer` anyway.  We also produce a `candle` definition for each bar (e.g. body size, etc.), so we could just return that if helpful, but it is perhaps more data than needed.

Here's a summary of things I'll explore:

- [x] common output format for candles (maybe a combination of interface and extended classes for multi-candle patterns)
- [x] `enum` or other return value to depict the match quality
- [x] metadata, including hit price, to support better characterization (maybe include candle data)
- [ ] ability to request multiple patterns in one call

Currently, the standard approach is to use a custom return class (e.g. `MarubozuResult`); however, I think we can either use a common base interface or extension classes.  I'm not 100% sure we'll be able to have a common `CandleResult` or `PatternResult` return class yet, but I'll look at some different options and see if that would be problematic or not.

**Reply by DaveSkender** [reply]
For the enum, here's what I'm thinking:

```csharp
public enum Signal
{
    BullConfirmed = 200,
    BullSignal = 100,
    BullBasis = 10,
    None = 0,
    BearBasis = -10,
    BearSignal = -100,
    BearConfirmed = -200
}
```
The `basis` variant will be used to paint the predecessor candles used to define the pattern signal.  For example, "three crows" is based on 3+ candles, so the first 2+ would be painted as `basis`, the last would be `signal`, and the following as `confirmed` where appropriate.  See #683

**Reply by jimtollan** [reply]
Dave - was pondering this as i lay in the bath (it's after midnight here in the UK!). do you think there's a *version* of these candles where each individual candle could be *aware* of its predecessor?? My thinking here is that by passing info using a chinese whisper type DNA approach to the candles, info about the state of formation could be passed across the candidates. In short, no two SAME looking candles would have the same score/weighting, as their current score would depend totally on their predecessor (who'd also have been signalled by their predecessor -etc, etc).

As there are basically only 5 main individual candle types, this might allow the more complex multi-candle configurations to be more loosely defined. In fact, there may end up being no need to define those explicitly as the *DNA* score on the grouping would signal a weighted value for the troupe (based on the history chain).

Anyway, bath water obviously flooded the brain via the ears - just a thought ;)

**Reply by jimtollan** [reply]
Dave - I was going to suggest pro-rating the OHLC values in the quote data, so that the CandleResult would return generically similar numbers whether loking at BTC or one of the alt coins. I did mention this concept before, but thought maybe a little extension method would make it easier for you to try with your code. Ignore the inefficient coding, trying to keep it as simple as possible, so as to demonstrate the pro-rating functionality

**Reply by DaveSkender** [reply]
> Dave - I was going to suggest pro-rating the OHLC values in the quote data, so that the CandleResult would return generically similar numbers whether looking at BTC or one of the alt coins. I did mention this concept before, but thought maybe a little extension method would make it easier for you to try with your code. Ignore the inefficient coding, trying to keep it as simple as possible, so as to demonstrate the pro-rating functionality

My view on this is that the library will take whatever quote history you provide.  If someone wants to alter their data before it goes in, they are welcome to do that.  I had assumed you were only doing this to get around some limitations on some other third party tool you were using.  Since we take `decimal` quote data, using the larger cryto values won't be an issue.

The concept you're describing is one of "normalizing" data, which I get, but am not understanding the value of adding something like that to this library + I'm not yet convinced it's a good idea (e.g. you'd be really messing with the actual rate of return values, for example).

**Reply by jimtollan** [reply]
Dave - points taken, was just my wandering (and fairly inexperienced in this field) mind!!

Without changing focus away from this too much. I wondered if you could point me to any supporting functions either present in Indicators, or planned that would allow one to identify either crossovers or crossunders on two series, such as tow EMA's etc. This would be a useful feature to have. Again, don't mean to pepper this thread with OT questions, but thought I might ask.

**Reply by DaveSkender** [reply]
This discussion is about candlestick patterns, so I’d suggest starting a new discussion or continuing an existing discussion on the topic.  Check #484

**Reply by DaveSkender** [reply]
I've made some updates in v1.21.0 for review, feedback is appreciated.

**Reply by jimtollan** [reply]
Cheers Dave - just catching up with the week!! Will have a look around and get back to you over the coming weeks. as ever, great to see this project fully alive and still coming up with novel ideas.

**Comment by jimtollan** [comment]
![image](https://user-images.githubusercontent.com/87867061/149957129-0ead73dc-641b-4931-bf67-e34761c0261b.png)

usage: `var proratedQuotes = quotes.ProRate();`
